distributed/scheduler.py

```diff
@@ -4229,19 +4235,17 @@ async def add_worker(
             client_msgs: dict = {}
             worker_msgs: dict = {}
             if nbytes:
+                assert isinstance(nbytes, dict)
                 for key in nbytes:
                     ts: TaskState = parent._tasks.get(key)
-                    if ts is not None and ts._state in ("processing", "waiting"):
-                        t: tuple = parent._transition(
-                            key,
-                            "memory",
+                    if ts is not None:
+                        self.handle_task_finished(
+                            key=key,
                             worker=address,
                             nbytes=nbytes[key],
                             typename=types[key],
+                            metadata=ts.metadata,
                         )
-                        recommendations, client_msgs, worker_msgs = t
-                        parent._transitions(recommendations, client_msgs, worker_msgs)
-                        recommendations = {}
```

Comment:
So the point of this was to collect all messages in one place before sending them out. If we are making this change, it would fragment communication

Reply:
Is your concern about performance or consistency?

I was mostly concerned about consistency and wanted to avoid code duplication. the communication would still happen via a `BatchedSend`. I would think that the performance concern of fragmented communication is something which is solved by the batched send but if you disagree I'd be curious to know why.

Reply:
We did a bunch of work to batch communication and improve general performance ( https://github.com/dask/distributed/pull/4451 ). It would be preferable to keep that if we can.

Every time we call `send` it [triggers a wakeup]( https://github.com/dask/distributed/blob/06a70aaf23d3c2858496e6a21a37a7b46f3a8498/distributed/batched.py#L140-L142 ) of [background send]( https://github.com/dask/distributed/blob/06a70aaf23d3c2858496e6a21a37a7b46f3a8498/distributed/batched.py#L78-L79 ), which will [cause communication to happen]( https://github.com/dask/distributed/blob/06a70aaf23d3c2858496e6a21a37a7b46f3a8498/distributed/batched.py#L93-L95 ) as long as [enough time has elapsed]( https://github.com/dask/distributed/blob/06a70aaf23d3c2858496e6a21a37a7b46f3a8498/distributed/batched.py#L86-L88 ). We have found this overall process to be quite expensive ( https://github.com/dask/distributed/issues/4454 )

Reply:
I rolled back the changes. I don't think performance in this case would be too hurtful, though. After all reconnecting workers is rather a rare edge case and not the norm. Still, the additional if statement is fine for me.